File: 290-word-pattern/solution.py
class Solution:
    def wordPattern(self, pattern, s_str):
        """
        :type pattern: str
        :type str: str
        :rtype: bool
        """
        s_str = s_str.split()
        m_map1 = HashMap()
        m_map2 = HashMap()
        if len(pattern) != len(s_str):
            return False
        for i, s in enumerate(pattern):
            if m_map1.put(s, i) != m_map2.put(s_str[i], i):
                return False
        return True

class HashMap(dict):
    def put(self, k, v):
        if k not in self:
            self[k] = v
            return None
        else:
            ret = self.get(k)
            self[k] = v
            return ret

File: 290-word-pattern/solution.cpp
class Solution {
public:
    bool wordPattern(string pattern, string s) {
        map<char, int> pIndexMap;
        map<string, int> wIndexMap;
        istringstream in(s);
        int i = 0, n = pattern.size();
        string word;
        while (in >> word) {
            if (i == n || pIndexMap[pattern[i]] != wIndexMap[word])
                return false;
            pIndexMap[pattern[i]] = wIndexMap[word] = i + 1;
            i ++;
        }
        return i == n;
    }
};
